test/integration/validation/chain_validator_storage_test.cpp

```diff
@@ -0,0 +1,162 @@
+          .quorum(1);
+    }
+
+    /// Complete builder by adding a signature and return a signed transaction
+    template <typename Builder>
+    auto completeTx(Builder builder) {
+      return builder.build().signAndAddSignature(keys.at(0)).finish();
+    }
+
+    /// Generate a dummy transaction with create role command
+    auto dummyTx(std::size_t i) {
+      return completeTx(baseTx().createRole("role" + std::to_string(i), {}));
+    }
+
+    /// Create block unsigned wrapper with given transactions, height, prev hash
+    auto baseBlock(std::vector<shared_model::proto::Transaction> transactions,
+                   shared_model::interface::types::HeightType height,
+                   shared_model::interface::types::HashType prev_hash) {
+      return shared_model::proto::BlockBuilder()
+          .transactions(transactions)
+          .height(height)
+          .prevHash(prev_hash)
+          .createdTime(iroha::time::now())
+          .build();
+    }
+
+    /// Complete wrapper and return a signed object
+    template <typename Wrapper>
+    auto completeBlock(Wrapper &&wrapper) {
+      return std::forward<Wrapper>(wrapper).finish();
+    }
+
+    /// Create mutable storage from initialized storage
+    auto createMutableStorage() {
+      return boost::get<
+                 expected::Value<std::unique_ptr<ametsuchi::MutableStorage>>>(
+                 storage->createMutableStorage())
+          .value;
+    }
+
+    /// Create first block with 4 peers, apply it to storage and return it
+    auto generateAndApplyFirstBlock() {
+      auto tx =
+          completeTx(baseTx()
+                         .addPeer("0.0.0.0:50541", keys.at(0).publicKey())
+                         .addPeer("0.0.0.0:50542", keys.at(1).publicKey())
+                         .addPeer("0.0.0.0:50543", keys.at(2).publicKey())
+                         .addPeer("0.0.0.0:50544", keys.at(3).publicKey()));
+
+      auto block = completeBlock(
+          baseBlock({tx},
+                    1,
+                    shared_model::crypto::DefaultHashProvider::makeHash(
+                        shared_model::crypto::Blob("")))
+              .signAndAddSignature(keys.at(0)));
+
+      auto ms = createMutableStorage();
+
+      ms->apply(block);
+      storage->commit(std::move(ms));
+
+      return block;
+    }
+
+    /// Create an observable from chain and return its validation status
+    auto createAndValidateChain(
+        std::vector<std::shared_ptr<shared_model::interface::Block>> chain) {
+      auto ms = createMutableStorage();
+      return validator->validateChain(rxcpp::observable<>::iterate(chain), *ms);
+    }
+
+    std::shared_ptr<validation::ChainValidatorImpl> validator;
+    std::vector<shared_model::crypto::Keypair> keys;
+  };
+
+  /**
+   * @given initialized storage with 4 peers, second block with add peer
+   * command, third block signed by new peer, blocks are signed by supermajority
+   * of ledger peers
```

Comment:
Telling honestly, I understood nothing from the raw description of test without looking to the code. Can you please reformulate it a bit, for example, by reflecting blocks states as elements of list?